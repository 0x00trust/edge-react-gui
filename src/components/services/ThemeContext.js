// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig.js'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
