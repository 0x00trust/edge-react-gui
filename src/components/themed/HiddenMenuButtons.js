// @flow

import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'
import { cacheStyles } from 'react-native-patina'

import T from '../../modules/UI/components/FormattedText/FormattedText.ui'
import { type Theme, useTheme } from '../services/ThemeContext.js'

type HiddenButtonProps = { text: string | React.Node, onPress: () => Promise<void>, type: 'danger' | 'success' | 'default' }

type Props = {
  rightSwipable?: HiddenButtonProps,
  leftSwipable?: HiddenButtonProps,
  right?: HiddenButtonProps,
  left?: HiddenButtonProps
}

export const HIDDEN_MENU_BUTTONS_WIDTH = 5.5

export function HiddenMenuButtons(props: Props) {
  const { rightSwipable, leftSwipable, left, right } = props
  const theme = useTheme()
  const styles = getStyles(theme)
  const rightButtonWidth = !right || !rightSwipable ? styles.fullWidthButton : null
  const leftButtonWidth = !left || !leftSwipable ? styles.fullWidthButton : null

  return (
    <View style={styles.background}>
      <View style={styles.buttons}>
        {left && (
          <TouchableOpacity style={[styles.button, styles[left.type], leftButtonWidth]} onPress={left.onPress}>
            {typeof left.text === 'string' ? <T style={styles.text}>{left.text}</T> : left.text}
          </TouchableOpacity>
        )}
        {leftSwipable && (
          <TouchableOpacity style={[styles.button, styles[leftSwipable.type], leftButtonWidth]} onPress={leftSwipable.onPress}>
            {typeof leftSwipable.text === 'string' ? <T style={styles.text}>{leftSwipable.text}</T> : leftSwipable.text}
          </TouchableOpacity>
        )}
      </View>
      <View style={styles.buttons}>
        {rightSwipable && (
          <TouchableOpacity style={[styles.button, styles[rightSwipable.type], rightButtonWidth]} onPress={rightSwipable.onPress}>
            {typeof rightSwipable.text === 'string' ? <T style={styles.text}>{rightSwipable.text}</T> : rightSwipable.text}
          </TouchableOpacity>
        )}
        {right && (
          <TouchableOpacity style={[styles.button, styles[right.type], rightButtonWidth]} onPress={right.onPress}>
            {typeof right.text === 'string' ? <T style={styles.text}>{right.text}</T> : right.text}
          </TouchableOpacity>
        )}
      </View>
    </View>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  buttons: {
    height: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: theme.rem(HIDDEN_MENU_BUTTONS_WIDTH)
  },
  button: {
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    width: theme.rem(HIDDEN_MENU_BUTTONS_WIDTH / 2)
  },
  fullWidthButton: {
    width: theme.rem(HIDDEN_MENU_BUTTONS_WIDTH)
  },
  danger: {
    backgroundColor: theme.sliderTabSend
  },
  default: {
    backgroundColor: theme.sliderTabMore
  },
  success: {
    backgroundColor: theme.sliderTabRequest
  },
  text: {
    color: theme.primaryText
  },
  background: {
    alignItems: 'center',
    backgroundColor: theme.secondaryButton,
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between'
  }
}))
