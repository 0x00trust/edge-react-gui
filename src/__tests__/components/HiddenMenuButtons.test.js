/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext.js'
import { HiddenMenuButtonsComponent as Request } from '../../components/themed/HiddenMenuButtons.js'

describe('Request', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      rightSwipable: {
        label: 'string',
        children: 'string',
        color: 'success',
        onPress: async () => undefined
      },
      leftSwipable: {
        label: 'string',
        children: 'string',
        color: 'success',
        onPress: async () => undefined
      },
      right: {
        label: 'string',
        children: 'string',
        color: 'success',
        onPress: async () => undefined
      },
      left: {
        label: 'string',
        children: 'string',
        color: 'success',
        onPress: async () => undefined
      },
      isSwipingRight: true,
      isSwipingLeft: true,
      swipeDirection: 'left',
      theme: getTheme()
    }
    const actual = renderer.render(<Request {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
