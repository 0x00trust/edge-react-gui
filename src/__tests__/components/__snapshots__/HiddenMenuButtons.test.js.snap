// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request should render with loading props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "height": "100%",
            "width": 70,
          },
          undefined,
        ]
      }
    >
      <SquareButton
        color="success"
        label="string"
        onPress={[Function]}
      >
        string
      </SquareButton>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#77C513",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "height": "100%",
              "width": 70,
            },
            false,
          ]
        }
      >
        <SquareButton
          color="success"
          label="string"
          onPress={[Function]}
        >
          string
        </SquareButton>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
          },
          Object {
            "alignItems": "flex-end",
          },
          Object {
            "backgroundColor": "#77C513",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "height": "100%",
              "width": 70,
            },
            false,
          ]
        }
      >
        <SquareButton
          color="success"
          label="string"
          onPress={[Function]}
        >
          string
        </SquareButton>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "height": "100%",
            "width": 70,
          },
          undefined,
        ]
      }
    >
      <SquareButton
        color="success"
        label="string"
        onPress={[Function]}
      >
        string
      </SquareButton>
    </View>
  </View>
</View>
`;
